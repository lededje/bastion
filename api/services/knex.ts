import knex from 'knex';

import config from '../config/knexfile';

const connection = knex(config.production);

export default connection;