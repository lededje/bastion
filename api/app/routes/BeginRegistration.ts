import ValidationErrorResponse from '../types/RequestBodyValidationErrorResponse';
import express, { RequestHandler } from 'express'
import { body, validationResult } from 'express-validator';

import knex from '../services/knex';
import mail from '../services/mail';

import randomWords, { words } from '../utils/randomWords';
import { AccessToken, User } from '../types/Tables';

const app = express();

const welcomeEmail = (name: string, securityPhrase: string[], requestToken: string) => [
  `Hello ${name},`,
  '',
  `Complete sign up by clicking this link: http://localhost:3000/api/authenticate/${requestToken}`,
  '',
  `Security phrase: ${securityPhrase.join(', ')}`,
  '',
  'If you did not sign up to Bastion, please click this link:',
  '',
  'Best,',
  '',
  'Bastion Team',
].join('\n');

export type BeginRegistrationRequestBody = {
  name: string;
  email: string;
}

export type BeginRegistrationResponse = {
  securityPhrase: (typeof words[number])[];
}

export type EmailExistsErrorResponse = {
  error: 'EMAIL_ALREADY_IN_USE',
  message: 'The email provided is already in use',
}

type BeginRegistrationResponses =
  | BeginRegistrationResponse
  | ValidationErrorResponse
  | EmailExistsErrorResponse;

const BeginRegistration:RequestHandler<{}, BeginRegistrationResponses, BeginRegistrationRequestBody> = async (
  req,
  res
) => {
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    res.status(400).json({
      error: 'REQUSET_BODY_VALIDATION_ERROR',
      message: 'The request contained invalid properties',
      context: errors.array(),
    });
    return;
  }

  const { email, name } = req.body;

  const securityPhrase = randomWords(4);

  let user: User, accessToken: AccessToken;

  try {
    const [newUser] = await knex('users')
      .insert({ name, email }, '*')

    user = newUser;

    const [newAccessToken] = await knex('access_tokens')
      .insert({ user_id: user.id }, '*')

    accessToken = newAccessToken;

  } catch (e) {
    if(e.constraint === 'users_email_unique') {
      res.status(400).json({
        error: 'EMAIL_ALREADY_IN_USE',
        message: 'The email provided is already in use',
      });
      return;
    }
    throw e
  }

  await mail.sendMail({
    from: 'Team Bastion <noreply@likeminded.io>',
    to: `${name} <${email}>`,
    subject: "Complete Bastion Registration",
    text: welcomeEmail(name, securityPhrase, accessToken.request_token)
  });

  res.json({
    securityPhrase,
  });
}

app.post('/', [
  body('email').isEmail(),
  body('name').isLength({ min: 2 }),
], BeginRegistration);

export default app;
