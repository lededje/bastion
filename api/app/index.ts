import express from 'express'

import Healthcheck from './routes/Healthcheck';
import BeginRegistration from './routes/BeginRegistration';
import CreateSession from './routes/CreateSession';
import ValidateSession from './routes/ValidateSession';
import AvailableNumbers from './routes/AvailableNumbers';

const app = express();

app.use(express.json());
app.use(express.urlencoded({ extended: true }));

app.use('/', Healthcheck);
app.use('/onboarding/begin-registration', BeginRegistration);
app.use('/authentication/create-session', CreateSession);
app.use('/authentication/validate-session', ValidateSession);
app.use('/available-numbers', AvailableNumbers);

export default app;