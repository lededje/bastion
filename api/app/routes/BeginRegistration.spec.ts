jest.mock('../middleware/Authenticated');

jest.mock('../services/mail');
jest.mock('../utils/randomWords');

const mail = jest.requireMock('../services/mail');

import request from 'supertest'
import knex from '../services/knex'

import app from '../';

import { BeginRegistrationResponses } from './BeginRegistration'

describe('BeginRegistration POST /', () => {
  beforeEach(() => {
    jest.spyOn(mail, 'sendMail')
  })

  afterEach(() => {
    jest.spyOn(mail, 'sendMail').mockClear();
  })

  it('errors when info is missing', (done) => {
    request(app)
      .post('/onboarding/begin-registration')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(400, (err, resp) => {
        if(err) done(err);
        // Expect an array of friendlyNames and phoneNumbers
        expect(resp.body).toEqual<BeginRegistrationResponses>(expect.objectContaining({
          error: 'REQUSET_BODY_VALIDATION_ERROR',
          message: 'The request contained invalid properties',    
        }))
        done();
      });
  });

  it('errors when email is already in use', async () => {
    const USER = { email: 'alreadyinuse@example.com', name: 'Alread In-Use' }

    const [existingUser] = await knex('users')
      .insert(USER)
      .returning('*');

    try {
      const response = await request(app)
        .post('/onboarding/begin-registration')
        .send(USER)
        .set('Accept', 'application/json')
        .set('Content-Type', 'application/json')
      
      expect(response.headers['content-type']).toEqual(expect.stringMatching(/json/));
      
      expect(response.body).toEqual<BeginRegistrationResponses>(({
        error: 'EMAIL_ALREADY_IN_USE',
        message: 'The email provided is already in use'
      }));

      expect(response.status).toEqual(400);
    } finally {
      await knex('users')
        .where({ id: existingUser.id })
        .delete();
    }
  });

  it('sends an email to the new user', async () => {
    const USER = { email: 'newuser@example.com', name: 'New User' };

    try {
      await request(app)
        .post('/onboarding/begin-registration')
        .send(USER)
        .set('accept', 'application/json')
        .set('content-type', 'application/json')
      
      const emailDescription = await (async () => {
        const rawMessage = await mail.sendMail.mock.results[0].value;
        const rawJson = JSON.parse(rawMessage.message);
        return {
          ...rawJson,
          messageId: 'Message ID removed for test consistency',
          text: rawJson.text.replace(/[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}/ig, 'removed-uuid-v4-for-tests')
        }
      })();

      expect(emailDescription).toMatchSnapshot();
    } finally {
      await knex('users')
        .where(USER)
        .delete();
    }
  });
});

afterAll(() => {
  knex.destroy()
})