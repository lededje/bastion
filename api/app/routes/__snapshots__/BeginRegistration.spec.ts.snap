// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeginRegistration POST / sends an email to the new user 1`] = `
Object {
  "from": Object {
    "address": "noreply@likeminded.io",
    "name": "Team Bastion",
  },
  "headers": Object {},
  "messageId": "Message ID removed for test consistency",
  "subject": "Complete Bastion Registration",
  "text": "Hello New User,

Complete sign up by clicking this link: http://localhost:3000/api/authenticate/removed-uuid-v4-for-tests

Security phrase: MOCKED_RANDOM_WORD, MOCKED_RANDOM_WORD, MOCKED_RANDOM_WORD, MOCKED_RANDOM_WORD

If you did not sign up to Bastion, please click this link:

Best,

Bastion Team",
  "to": Array [
    Object {
      "address": "newuser@example.com",
      "name": "New User",
    },
  ],
}
`;
